core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,221 @@
+  private val MAX_FAILED_EXEC_PER_NODE = conf.get(config.MAX_FAILED_EXEC_PER_NODE)
+  val EXECUTOR_RECOVERY_MILLIS = BlacklistTracker.getBlacklistExpiryTime(conf)
+
+  // a count of failed tasks for each executor.  Only counts failures after tasksets complete
+  // successfully
+  private val executorIdToFailureCount: HashMap[String, Int] = HashMap()
+  private val executorIdToBlacklistExpiryTime: HashMap[String, Long] = new HashMap()
+  private val nodeIdToBlacklistExpiryTime: HashMap[String, Long] = new HashMap()
+  private val _nodeBlacklist: AtomicReference[Set[String]] = new AtomicReference(Set())
+  private var nextExpiryTime: Long = Long.MaxValue
+
+  def start(): Unit = {}
+
+  def stop(): Unit = {}
+
+  def expireExecutorsInBlacklist(): Unit = {
+    val now = clock.getTimeMillis()
+    // quickly check if we've got anything to expire from blacklist -- if not, avoid doing any work
+    if (now > nextExpiryTime) {
+      val execsToClear = executorIdToBlacklistExpiryTime.filter(_._2 < now).keys
+      if (execsToClear.nonEmpty) {
+        logInfo(s"Removing executors $execsToClear from blacklist during periodic recovery")
+        execsToClear.foreach { exec => executorIdToBlacklistExpiryTime.remove(exec) }
+      }
+      if (executorIdToBlacklistExpiryTime.nonEmpty) {
+        nextExpiryTime = executorIdToBlacklistExpiryTime.map{_._2}.min
+      } else {
+        nextExpiryTime = Long.MaxValue
+      }
+      val nodesToClear = nodeIdToBlacklistExpiryTime.filter(_._2 < now).keys
+      if (nodesToClear.nonEmpty) {
+        logInfo(s"Removing nodes $nodesToClear from blacklist during periodic recovery")
+        nodesToClear.foreach { node => nodeIdToBlacklistExpiryTime.remove(node) }
+        // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+        _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+      }
+    }
+ }
+
+  def taskSetSucceeded(
+      failuresByExec: HashMap[String, FailureStatus],
+      scheduler: TaskSchedulerImpl): Unit = {
+    // if any tasks failed, we count them towards the overall failure count for the executor at
+    // this point.
+    failuresByExec.foreach { case (exec, newFailures) =>
+      val prevFailures = executorIdToFailureCount.getOrElse(exec, 0)
+      val newTotal = prevFailures + newFailures.totalFailures
+
+      if (newTotal >= MAX_FAILURES_PER_EXEC) {
+        logInfo(s"Blacklisting executor $exec because it has $newTotal" +
+          s" task failures in successful task sets")
+        val now = clock.getTimeMillis()
+        val expiryTime = now + EXECUTOR_RECOVERY_MILLIS
+        executorIdToBlacklistExpiryTime.put(exec, expiryTime)
+        executorIdToFailureCount.remove(exec)
+        if (expiryTime < nextExpiryTime) {
+          nextExpiryTime = expiryTime
+        }
+
+        val node = scheduler.getHostForExecutor(exec)
+        val execs = scheduler.getExecutorsAliveOnHost(node).getOrElse(Set())
```

Comment:
oh good point.  this design was somewhat intentional -- to deal w/ intermittent flakiness, you wouldn't blacklist a node if eg. you had one executor get blacklisted on it 3 times, but with long periods in between.  I see though that this is probably undesirable with DA.

Its actually a really good point, that with DA node blacklisting is important even with just one executor per node.

We could either:
1) keep it as it is now, so each new executor that DA adds has to get re-blacklisted. 
2) always blacklist the node after N blacklisted executors, regardless of the time between
3) add a timeout for how much time can go between blacklisted executors for it to count.  Then we've got to either add another configuration or re-use one of the existing ones.

I don't like the added configuration burden of 3 ... I'm leaning towards 2.  Though maybe I'm forced to use 3 just because otherwise we track failed executors for all time, which I guess would be a memory leak, though hopefully an extremely slow one.